e2e/pos_test.go

```diff
@@ -82,6 +83,119 @@ func validateValidatorSet(
 	}
 }
 
+func TestPoS_ValidatorBoundaries(t *testing.T) {
+	accounts := []struct {
+		key     *ecdsa.PrivateKey
+		address types.Address
+	}{}
+	stakeAmount := framework.EthToWei(1)
+	numGenesisValidators := IBFTMinNodes
+	minValidatorCount := uint32(1)
+	maxValidatorCount := uint32(numGenesisValidators + 1)
+	numNewStakers := 2
+
+	for i := 0; i < numNewStakers; i++ {
+		k, a := tests.GenerateKeyAndAddr(t)
+
+		accounts = append(accounts, struct {
+			key     *ecdsa.PrivateKey
+			address types.Address
+		}{
+			key:     k,
+			address: a,
+		})
+	}
+
+	defaultBalance := framework.EthToWei(100)
+	ibftManager := framework.NewIBFTServersManager(
+		t,
+		numGenesisValidators,
+		IBFTDirPrefix,
+		func(i int, config *framework.TestServerConfig) {
+			config.SetSeal(true)
+			config.SetEpochSize(2)
+			config.PremineValidatorBalance(defaultBalance)
+			for j := 0; j < numNewStakers; j++ {
+				config.Premine(accounts[j].address, defaultBalance)
+			}
+			config.SetIBFTPoS(true)
+			config.SetMinValidatorCount(minValidatorCount)
+			config.SetMaxValidatorCount(maxValidatorCount)
+		})
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	ibftManager.StartServers(ctx)
+
+	srv := ibftManager.GetServer(0)
+
+	client := srv.JSONRPC()
+
+	testCases := []struct {
+		name              string
+		address           types.Address
+		key               *ecdsa.PrivateKey
+		expectedExistence bool
+		expectedSize      int
+	}{
+		{
+			name:              "Can add a 5th validator",
+			address:           accounts[0].address,
+			key:               accounts[0].key,
+			expectedExistence: true,
+			expectedSize:      numGenesisValidators + 1,
+		},
+		{
+			name:              "Can not add a 6th validator",
+			address:           accounts[1].address,
+			key:               accounts[1].key,
+			expectedExistence: false,
+			expectedSize:      numGenesisValidators + 1,
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			err := framework.StakeAmount(tt.address, tt.key, stakeAmount, srv)
+			assert.NoError(t, err)
```

Comment:
Why don't you verify that the transaction to stake actually errored out with the appropriate SC reason?

You just send a transaction here, don't even wait for a receipt - just check the validator set

Reply:
No sure i understand. I know that framework.StakeAmount waits for receipt. Should i just send regular transaction or modify framework.StakeAmount not to wait for receipt

Reply:
Resolving as we don't have a normal testing framework, but some clunky mechanism that just sends transactions. This fix is out of the scope of this PR